File: java/lib/src/main/java/com/svix/models/StreamSinkOutConfig.java
// This file is @generated
package com.svix.models;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.svix.Utils;

import lombok.*;

import java.util.HashMap;
import java.util.Map;

@ToString
@EqualsAndHashCode
public abstract class StreamSinkOutConfig {
    @JsonIgnore
    public String getVariantName() {
        VariantName annotation = this.getClass().getAnnotation(VariantName.class);
        return annotation != null ? annotation.value() : null;
    }

    public abstract JsonNode toJsonNode();

    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("poller")
    public static class Poller extends StreamSinkOutConfig {
        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().createObjectNode();
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("azureBlobStorage")
    public static class AzureBlobStorage extends StreamSinkOutConfig {
        private final AzureBlobStorageConfigOut azureBlobStorage;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(azureBlobStorage);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("otelTracing")
    public static class OtelTracing extends StreamSinkOutConfig {
        private final SinkOtelTracingConfigOut otelTracing;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(otelTracing);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("http")
    public static class Http extends StreamSinkOutConfig {
        private final SinkHttpConfigOut http;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(http);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("amazonS3")
    public static class AmazonS3 extends StreamSinkOutConfig {
        private final S3ConfigOut amazonS3;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(amazonS3);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("snowflake")
    public static class Snowflake extends StreamSinkOutConfig {
        private final SnowflakeConfigOut snowflake;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(snowflake);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("googleCloudStorage")
    public static class GoogleCloudStorage extends StreamSinkOutConfig {
        private final GoogleCloudStorageConfigOut googleCloudStorage;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(googleCloudStorage);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("googleCloudPubSub")
    public static class GoogleCloudPubSub extends StreamSinkOutConfig {
        private final GoogleCloudPubSubConfigOut googleCloudPubSub;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(googleCloudPubSub);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("redshift")
    public static class Redshift extends StreamSinkOutConfig {
        private final RedshiftConfigOut redshift;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(redshift);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("bigQuery")
    public static class BigQuery extends StreamSinkOutConfig {
        private final BigQueryConfigOut bigQuery;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(bigQuery);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("clickhouse")
    public static class Clickhouse extends StreamSinkOutConfig {
        private final ClickhouseConfigOut clickhouse;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(clickhouse);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("rabbitMq")
    public static class RabbitMq extends StreamSinkOutConfig {
        private final RabbitMqConfigOut rabbitMq;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(rabbitMq);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("sqs")
    public static class Sqs extends StreamSinkOutConfig {
        private final SqsConfigOut sqs;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(sqs);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("eventBridge")
    public static class EventBridge extends StreamSinkOutConfig {
        private final EventBridgeConfigOut eventBridge;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(eventBridge);
        }
    }

    @Getter
    @Setter
    @AllArgsConstructor
    @ToString
    @EqualsAndHashCode(callSuper = false)
    @VariantName("sns")
    public static class Sns extends StreamSinkOutConfig {
        private final SnsConfigOut sns;

        @Override
        public JsonNode toJsonNode() {
            return Utils.getObjectMapper().valueToTree(sns);
        }
    }

    @FunctionalInterface
    private interface TypeFactory {
        StreamSinkOutConfig create(JsonNode config);
    }

    private static final Map<String, TypeFactory> TY_M = new HashMap<>();
    private static final ObjectMapper m = Utils.getObjectMapper();

    static {
        TY_M.put("poller", c -> new Poller());
        TY_M.put(
                "azureBlobStorage",
                c -> new AzureBlobStorage(m.convertValue(c, AzureBlobStorageConfigOut.class)));
        TY_M.put(
                "otelTracing",
                c -> new OtelTracing(m.convertValue(c, SinkOtelTracingConfigOut.class)));
        TY_M.put("http", c -> new Http(m.convertValue(c, SinkHttpConfigOut.class)));
        TY_M.put("amazonS3", c -> new AmazonS3(m.convertValue(c, S3ConfigOut.class)));
        TY_M.put("snowflake", c -> new Snowflake(m.convertValue(c, SnowflakeConfigOut.class)));
        TY_M.put(
                "googleCloudStorage",
                c -> new GoogleCloudStorage(m.convertValue(c, GoogleCloudStorageConfigOut.class)));
        TY_M.put(
                "googleCloudPubSub",
                c -> new GoogleCloudPubSub(m.convertValue(c, GoogleCloudPubSubConfigOut.class)));
        TY_M.put("redshift", c -> new Redshift(m.convertValue(c, RedshiftConfigOut.class)));
        TY_M.put("bigQuery", c -> new BigQuery(m.convertValue(c, BigQueryConfigOut.class)));
        TY_M.put("clickhouse", c -> new Clickhouse(m.convertValue(c, ClickhouseConfigOut.class)));
        TY_M.put("rabbitMq", c -> new RabbitMq(m.convertValue(c, RabbitMqConfigOut.class)));
        TY_M.put("sqs", c -> new Sqs(m.convertValue(c, SqsConfigOut.class)));
        TY_M.put(
                "eventBridge", c -> new EventBridge(m.convertValue(c, EventBridgeConfigOut.class)));
        TY_M.put("sns", c -> new Sns(m.convertValue(c, SnsConfigOut.class)));
    }

    public static StreamSinkOutConfig fromTypeAndConfig(String type, JsonNode config) {
        TypeFactory factory = TY_M.get(type);
        if (factory == null) {
            throw new IllegalArgumentException("Unknown type: " + type);
        }
        return factory.create(config);
    }
}

File: php/src/Models/StreamSinkOutConfig.php
<?php

// this file is @generated
declare(strict_types=1);

namespace Svix\Models;

abstract class StreamSinkOutConfig implements \JsonSerializable
{
    abstract public function variantName(): string;

    abstract public function configPayload(): mixed;

    public static function poller(): StreamSinkOutConfig\Poller
    {
        return new StreamSinkOutConfig\Poller();
    }

    public static function pollingEndpoint(): StreamSinkOutConfig\PollingEndpoint
    {
        return new StreamSinkOutConfig\PollingEndpoint();
    }

    public static function azureBlobStorage(AzureBlobStorageConfigOut $azureBlobStorage): StreamSinkOutConfig\AzureBlobStorage
    {
        return new StreamSinkOutConfig\AzureBlobStorage($azureBlobStorage);
    }

    public static function otelTracing(OtelTracingConfigOut $otelTracing): StreamSinkOutConfig\OtelTracing
    {
        return new StreamSinkOutConfig\OtelTracing($otelTracing);
    }

    public static function http(SinkHttpConfigOut $http): StreamSinkOutConfig\Http
    {
        return new StreamSinkOutConfig\Http($http);
    }

    public static function amazonS3(S3ConfigOut $amazonS3): StreamSinkOutConfig\AmazonS3
    {
        return new StreamSinkOutConfig\AmazonS3($amazonS3);
    }

    public static function snowflake(SnowflakeConfigOut $snowflake): StreamSinkOutConfig\Snowflake
    {
        return new StreamSinkOutConfig\Snowflake($snowflake);
    }

    public static function googleCloudStorage(GoogleCloudStorageConfigOut $googleCloudStorage): StreamSinkOutConfig\GoogleCloudStorage
    {
        return new StreamSinkOutConfig\GoogleCloudStorage($googleCloudStorage);
    }

    public static function googleCloudPubSub(GoogleCloudPubSubConfigOut $googleCloudPubSub): StreamSinkOutConfig\GoogleCloudPubSub
    {
        return new StreamSinkOutConfig\GoogleCloudPubSub($googleCloudPubSub);
    }

    public static function redshift(RedshiftConfigOut $redshift): StreamSinkOutConfig\Redshift
    {
        return new StreamSinkOutConfig\Redshift($redshift);
    }

    public static function bigQuery(BigQueryConfigOut $bigQuery): StreamSinkOutConfig\BigQuery
    {
        return new StreamSinkOutConfig\BigQuery($bigQuery);
    }

    public static function clickhouse(ClickhouseConfigOut $clickhouse): StreamSinkOutConfig\Clickhouse
    {
        return new StreamSinkOutConfig\Clickhouse($clickhouse);
    }

    public static function rabbitMq(RabbitMqConfigOut $rabbitMq): StreamSinkOutConfig\RabbitMq
    {
        return new StreamSinkOutConfig\RabbitMq($rabbitMq);
    }

    public static function sqs(SqsConfigOut $sqs): StreamSinkOutConfig\Sqs
    {
        return new StreamSinkOutConfig\Sqs($sqs);
    }

    public static function eventBridge(EventBridgeConfigOut $eventBridge): StreamSinkOutConfig\EventBridge
    {
        return new StreamSinkOutConfig\EventBridge($eventBridge);
    }

    public static function sns(SnsConfigOut $sns): StreamSinkOutConfig\Sns
    {
        return new StreamSinkOutConfig\Sns($sns);
    }

    public static function postgres(PostgresConfigOut $postgres): StreamSinkOutConfig\Postgres
    {
        return new StreamSinkOutConfig\Postgres($postgres);
    }

    public static function fromTypeAndConfig(string $type, mixed $config): self
    {
        $config ??= [];

        return match ($type) {
            'poller' => self::poller(),

            'pollingEndpoint' => self::pollingEndpoint(),

            'azureBlobStorage' => self::azureBlobStorage(AzureBlobStorageConfigOut::fromMixed($config)),

            'otelTracing' => self::otelTracing(OtelTracingConfigOut::fromMixed($config)),

            'http' => self::http(SinkHttpConfigOut::fromMixed($config)),

            'amazonS3' => self::amazonS3(S3ConfigOut::fromMixed($config)),

            'snowflake' => self::snowflake(SnowflakeConfigOut::fromMixed($config)),

            'googleCloudStorage' => self::googleCloudStorage(GoogleCloudStorageConfigOut::fromMixed($config)),

            'googleCloudPubSub' => self::googleCloudPubSub(GoogleCloudPubSubConfigOut::fromMixed($config)),

            'redshift' => self::redshift(RedshiftConfigOut::fromMixed($config)),

            'bigQuery' => self::bigQuery(BigQueryConfigOut::fromMixed($config)),

            'clickhouse' => self::clickhouse(ClickhouseConfigOut::fromMixed($config)),

            'rabbitMq' => self::rabbitMq(RabbitMqConfigOut::fromMixed($config)),

            'sqs' => self::sqs(SqsConfigOut::fromMixed($config)),

            'eventBridge' => self::eventBridge(EventBridgeConfigOut::fromMixed($config)),

            'sns' => self::sns(SnsConfigOut::fromMixed($config)),

            'postgres' => self::postgres(PostgresConfigOut::fromMixed($config)),

            default => throw new \InvalidArgumentException("Unknown type: {$type}"),
        };
    }

    /**
     * Create an instance from a mixed obj.
     */
    public static function fromMixed(mixed $data): self
    {
        return self::fromTypeAndConfig(
            \Svix\Utils::deserializeString($data, 'type', true, 'StreamSinkOutConfig'),
            \Svix\Utils::getValFromJson($data, 'config', false, 'StreamSinkOutConfig') ?? []
        );
    }

    /**
     * Create an instance from a json string.
     */
    public static function fromJson(string $json): self
    {
        $data = json_decode(json: $json, associative: true, depth: 512, flags: JSON_THROW_ON_ERROR);

        return self::fromMixed($data);
    }

    public function jsonSerialize(): mixed
    {
        return \Svix\Utils::newStdClassIfArrayIsEmpty($this->configPayload());
    }
}

namespace Svix\Models\StreamSinkOutConfig;

final class Poller extends \Svix\Models\StreamSinkOutConfig
{
    public function variantName(): string
    {
        return 'poller';
    }

    public function configPayload(): mixed
    {
        return new \stdClass();
    }
}

final class PollingEndpoint extends \Svix\Models\StreamSinkOutConfig
{
    public function variantName(): string
    {
        return 'pollingEndpoint';
    }

    public function configPayload(): mixed
    {
        return new \stdClass();
    }
}

final class AzureBlobStorage extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\AzureBlobStorageConfigOut $azureBlobStorage)
    {
    }

    public function variantName(): string
    {
        return 'azureBlobStorage';
    }

    public function configPayload(): mixed
    {
        return $this->azureBlobStorage;
    }
}

final class OtelTracing extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\OtelTracingConfigOut $otelTracing)
    {
    }

    public function variantName(): string
    {
        return 'otelTracing';
    }

    public function configPayload(): mixed
    {
        return $this->otelTracing;
    }
}

final class Http extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\SinkHttpConfigOut $http)
    {
    }

    public function variantName(): string
    {
        return 'http';
    }

    public function configPayload(): mixed
    {
        return $this->http;
    }
}

final class AmazonS3 extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\S3ConfigOut $amazonS3)
    {
    }

    public function variantName(): string
    {
        return 'amazonS3';
    }

    public function configPayload(): mixed
    {
        return $this->amazonS3;
    }
}

final class Snowflake extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\SnowflakeConfigOut $snowflake)
    {
    }

    public function variantName(): string
    {
        return 'snowflake';
    }

    public function configPayload(): mixed
    {
        return $this->snowflake;
    }
}

final class GoogleCloudStorage extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\GoogleCloudStorageConfigOut $googleCloudStorage)
    {
    }

    public function variantName(): string
    {
        return 'googleCloudStorage';
    }

    public function configPayload(): mixed
    {
        return $this->googleCloudStorage;
    }
}

final class GoogleCloudPubSub extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\GoogleCloudPubSubConfigOut $googleCloudPubSub)
    {
    }

    public function variantName(): string
    {
        return 'googleCloudPubSub';
    }

    public function configPayload(): mixed
    {
        return $this->googleCloudPubSub;
    }
}

final class Redshift extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\RedshiftConfigOut $redshift)
    {
    }

    public function variantName(): string
    {
        return 'redshift';
    }

    public function configPayload(): mixed
    {
        return $this->redshift;
    }
}

final class BigQuery extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\BigQueryConfigOut $bigQuery)
    {
    }

    public function variantName(): string
    {
        return 'bigQuery';
    }

    public function configPayload(): mixed
    {
        return $this->bigQuery;
    }
}

final class Clickhouse extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\ClickhouseConfigOut $clickhouse)
    {
    }

    public function variantName(): string
    {
        return 'clickhouse';
    }

    public function configPayload(): mixed
    {
        return $this->clickhouse;
    }
}

final class RabbitMq extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\RabbitMqConfigOut $rabbitMq)
    {
    }

    public function variantName(): string
    {
        return 'rabbitMq';
    }

    public function configPayload(): mixed
    {
        return $this->rabbitMq;
    }
}

final class Sqs extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\SqsConfigOut $sqs)
    {
    }

    public function variantName(): string
    {
        return 'sqs';
    }

    public function configPayload(): mixed
    {
        return $this->sqs;
    }
}

final class EventBridge extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\EventBridgeConfigOut $eventBridge)
    {
    }

    public function variantName(): string
    {
        return 'eventBridge';
    }

    public function configPayload(): mixed
    {
        return $this->eventBridge;
    }
}

final class Sns extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\SnsConfigOut $sns)
    {
    }

    public function variantName(): string
    {
        return 'sns';
    }

    public function configPayload(): mixed
    {
        return $this->sns;
    }
}

final class Postgres extends \Svix\Models\StreamSinkOutConfig
{
    public function __construct(public readonly \Svix\Models\PostgresConfigOut $postgres)
    {
    }

    public function variantName(): string
    {
        return 'postgres';
    }

    public function configPayload(): mixed
    {
        return $this->postgres;
    }
}
